### PR TITLE
Coupling Merra2 aerosol climatology and GOCART forecasted aerosol with Thompson microphysics

### PR DESCRIPTION
**Pull request by ChunxiZhang-NOAA**
All credit for code changes and description to Anning Cheng.

MERRA2 climatological aerosols are used to calculate number concentrations of ice friendly and liquid friendly aerosols. IN and CCN can be activated by using aerosols. The indirect interactions of aerosols and clouds are expected to happen. This PR is created to replace the NCAR/ccpp-physics PR#[720](https://github.com/NCAR/ccpp-physics/pull/720), which was created by Anning Cheng.  

The MERRA2 coupled Thompson (EXP MRAERO) shows improvement from the control run and the forecasted NIFA and NWFA approach (LTAERO), specially decreased cloud water bias as shown in the attached poster. More comprehensive comparison located at https://www.emc.ncep.noaa.gov/gmb/acheng/mraero/.

Because the GOCART coupled Thompson scheme shares the same IN/CCN activation module as the MERRA2 coupled Thompson scheme, we expect more improvements. Some results can be found here:
[gass_3_v1.pdf](https://github.com/ufs-community/ccpp-physics/files/9586872/gass_3_v1.pdf)

### CODE REVIEW

**Comment by grantfirl** [comment]
This PR is a little difficult to review because the ufs/dev branch is behind NCAR/main and yet it looks like this PR branch is including the latest changes in NCAR/main. In order to effectively review, it seems like we either need to 1) pull the latest NCAR/main into ufs/dev or 2) recreate the changes off the top of ufs/dev.

**Comment by ChunxiZhang-NOAA** [comment]
@grantfirl Yes, it is difficult to review it.  No more people are invited to review it for now. I will sync this branch with the ufs/dev branch after ufs/dev merged with NCAR/main.  Hopefully it will follow this schedule: the ufs/dev merges with NCAR/main after Dustin's PR;  fv3atm points ccpp-physics to ufs-community/ccpp-physics ufs/dev branch;  the PRs in ufs-community/ccpp-physics start.

**Comment by grantfirl** [comment]
@ChunxiZhang-NOAA I don't see upstream PRs for fv3atm or ufs-weather-model. It looks like you're at least adding one new variable that needs to be allocated by the host, right (mraerosol)? So, there should be changes in GFS_typedefs.

File: physics/GFS_PBL_generic_pre.meta
Comment (grantfirl) [inline physics/GFS_PBL_generic_pre.meta]: I realize that maybe this functionality only works for Thompson right now, but is it conceivable that it could be used with other schemes? I would avoid including `for_thompson` unless absolutely necessary. Also, I think that the latest standard name rules prefer `do_X` over `flag_for_X`.

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (grantfirl) [inline physics/GFS_rrtmgp_thompsonmp_pre.F90]: It isn't required to have empty init/run/finalize anymore, so this can be removed.

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (grantfirl) [inline physics/GFS_rrtmgp_thompsonmp_pre.F90]: See comment below.

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (dustinswales) [inline physics/GFS_rrtmgp_thompsonmp_pre.F90]: @ChunxiZhang-NOAA 
GFS_rrtmgp_thompsonmp.F90 has been replaced with GFS_rrtmgp_cloud_mp.F90.

It looks like the new functionality added alongside "mraerosol" was added to both  GFS_rrtmgp_cloud_mp.F90 and GFS_rrtmgp_thompsonmp.F90. Since GFS_rrtmgp_thompsonmp.meta wasn't modified, I'm assume that you're not using this scheme and we can remove it from here?

File: physics/GFS_PBL_generic_pre.meta
Comment (ChunxiZhang-NOAA) [reply physics/GFS_PBL_generic_pre.meta]: I believe it could be used with other schemes in the future. I will change the name to "do_merra2_aerosol_awareness"

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (ChunxiZhang-NOAA) [reply physics/GFS_rrtmgp_thompsonmp_pre.F90]: I don't think GFS_rrtmgp_thompsonmp_pre.F90 is used. I will delete this file.

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (ChunxiZhang-NOAA) [reply physics/GFS_rrtmgp_thompsonmp_pre.F90]: This file is not used. I will delete this file.

File: physics/GFS_rrtmgp_thompsonmp_pre.F90
Comment (ChunxiZhang-NOAA) [reply physics/GFS_rrtmgp_thompsonmp_pre.F90]: This file is not used. I will delete this file.

**Comment by grantfirl** [comment]
@ChunxiZhang-NOAA I reviewed this and will approve once the GFS_rrtmgp_thompson file is removed.

**Comment by ChunxiZhang-NOAA** [comment]
@AnningCheng-NOAA @gthompsnWRF Please review this PR at your convenience.

**Review by AnningCheng-NOAA — APPROVED** [review]
look good to me. Approved

**Comment by AnningCheng-NOAA** [comment]
looks good and just approved.

**Review by Qingfu-Liu — APPROVED** [review]
Although I did not run the tests, but the code changes look good to me.

**Comment by ChunxiZhang-NOAA** [comment]
@grantfirl @dustinswales Could you please review this PR again, thanks.

**Review by grantfirl — APPROVED** [review]

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [inline physics/module_mp_thompson.F90]: I believe you should put the couple lines of this IF-block directly into the IF-block that comes just prior to it.  Seems a waste to double-IF-Block like this.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [inline physics/module_mp_thompson.F90:2208]: This is NOT needed.  Line#1432 takes care of initializing the nc1d array which will then get the proper value in the actual calculation.  This clearly got introduced by someone else along the way but this is not needed so the IF-statement should be removed.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [inline physics/module_mp_thompson.F90:3390]: Similar to lline#2217, this IF-test should not be present here.  The initialization of the nc1d array in the driver routine should take proper care to give the value of Nt_c in the non-aerosol-aware scheme.  Not sure how that line got added, but it should not be necessary.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [inline physics/module_mp_thompson.F90:3631]: Same as two earlier comments.  These IF-tests should be eliminated please.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [inline physics/module_mp_thompson.F90:5781]: Same as three times earlier.  This IF-test is not needed and should be eliminated.

File: physics/mp_thompson.F90
Comment (gthompsnWRF) [inline physics/mp_thompson.F90:881]: **PLEASE provide a full description** of what is done inside this subroutine.  This is a BIG deal and a bunch of numbers are shown but no explanations for anything.  In particular, I wish to see what atmospheric constituents comprise the ```aerfld``` array index going from 1 to 15 and the hard-wired numbers are some sort of mass conversions probably.  But a solid description really must be included here.

File: physics/mynnedmf_wrapper.F90
Comment (gthompsnWRF) [inline physics/mynnedmf_wrapper.F90:500]: Similarly to prior comment... This is setting aerosol numbers to zero for mraerosol, but aren't aerosol numbers getting mixed when using the earlier ```ltaerosol``` setting?  I am not understanding these apparent inconsistencies.

File: physics/mynnedmf_wrapper.F90
Comment (gthompsnWRF) [inline physics/mynnedmf_wrapper.F90:490]: I readily admit that I do not understand what is being attempted in the PBL scheme, but I have my guess.  I assume that variables marked with value of 2 are getting PBL tendencies for mixing.  As such, I think that means that mixing ratios of cloud water and cloud ice are getting turbulent mixing.  Yet why not number concentrations?

**Review by gthompsnWRF — CHANGES_REQUESTED** [review]

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:1186]: Agreed. I will delete it.

File: physics/mynnedmf_wrapper.F90
Comment (ChunxiZhang-NOAA) [reply physics/mynnedmf_wrapper.F90:490]: Those look like a deprecated feature. I didn't see the code actually used it. Maybe just a placeholder.

**Review by dustinswales — APPROVED** [review]

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90]: Anning and I believe that those lines are necessary.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:2208]: We just follow the previous code for ltaerosol. If you want to delete it please create a separate PR to clean the code in the future.

File: physics/mynnedmf_wrapper.F90
Comment (ChunxiZhang-NOAA) [reply physics/mynnedmf_wrapper.F90:500]: qnwfa and qnifa are diagnosed from merra2. The vertical mixing is not needed.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [reply physics/module_mp_thompson.F90]: I just noticed that one IF-test is ```errflg``` and the other is ```errmsg```.  Is there really a case when one can be present and not the other?  Could you simply combine these 2 if-tests into a single one using something like ```if (present(errmsg) .and. present(errflg)) then```?

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [reply physics/module_mp_thompson.F90:2208]: It seems unnecessary to me do this in its own PR.  I'm asking for @grantfirl to weigh in for whether this is OK to do as part of this PR.  It just takes more time later to deal with a set of four one-liners.

File: physics/mp_thompson.F90
Comment (gthompsnWRF) [inline physics/mp_thompson.F90:280]: A new comment here about having all the other where statements using floating point 0.0 but the qc test is using simply the number zero without a floating point representation.  Shouldn't they all be consistent?  It also concerns me that this setting to zero ONLY appears to occur for MERRA aerosol so why is it not done for the previous?

**Review by gthompsnWRF — CHANGES_REQUESTED** [review]

File: physics/mp_thompson.F90
Comment (AnningCheng-NOAA) [reply physics/mp_thompson.F90:881]: In GOCART and MERRA2, aerosol are given as mixing ratio (kg/kg).  To convert from kg/kg to #/kg, the "unit mass" (mass of one particle) within the mass bins is calculated. A lognormal size distribution within aerosol bins is used to find the size based upon the median mass. NIFA is mainly summarized over five dust bins and NWFA over the other 10 bins. The parameters besides each bins are carefully tuned for a good performance of  the scheme.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90]: Usually errmsg and errflag present as a pair. But I'd leave the current code as is.

Comment (grantfirl) [reply physics/module_mp_thompson.F90]: Good point, @gthompsnWRF. I think that I agree with you. Both errflg and errmsg are the CCPP error-handling variables. They should either be both present or neither present. I'm guessing what is really intended for in the error-checking here is to do one thing if this scheme is being called from within CCPP and another when it is not. So, to me, it is cleaner to do like @gthompsnWRF suggests: 
`if (present(errmsg) .and. present(errflg)) then` 
`fill errmsg, set errflg, return` 
`else write standard out and stop`.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:2208]: I just did a regression test by deleting those lines. It changed baseline results.

File: physics/module_mp_thompson.F90
Comment (dustinswales) [reply physics/module_mp_thompson.F90:5781]: @ChunxiZhang-NOAA I propose we eliminate these if-statements since they weren't introduced by the developer, and our redundant.

File: physics/module_mp_thompson.F90
Comment (gthompsnWRF) [reply physics/module_mp_thompson.F90:2208]: If you look at my original line of code, Nt_c is divided by air density and these lines do not divide by air density.  Perhaps that is the cause.  I kinda put the burden of proof that these lines are needed on someone at EMC.  I believe I coded things correctly to work without this addition.  I went back and checked using Git's "Blame" feature and all I can say is that it was ported 4 years ago by Dom H. and he must have decided to introduce those lines.  They should be unnecessary.

Comment (grantfirl) [reply physics/module_mp_thompson.F90:2208]: To me, it looks like `nc1d` is certainly initialized previously, but it also looks like the preceding code has the possibility to modify `nc` from its initial value. I interpret this line as being protection against modification of `nc` when not using aerosol awareness. This begs the question of why isn't this IF-statement placed higher up if we're just going to reset it down here, but perhaps values of the other variables need a temporarily modified `nc` to calculate, e.g. `nu_c`,`lamc`, etc before `nc` is forced back to its initial value? I'm certainly not an expert in this code and @gthompsnWRF is. My opinion, given @gthompsnWRF expertise, is to make the changes he suggests prior to merging rather than a follow-up UNLESS all of the UFS regression testing has already been finished, no other changes based on Greg's review will change the results, AND it is time-sensitive to get this in. Otherwise, follow-up PRs just kinda add more work over getting the code "right" the first time.

Comment (grantfirl) [reply physics/module_mp_thompson.F90:2208]: I was also going to mention Greg's point that this code has been there since Dom added it to CCPP. Either this line existed from the antecedent codebase he was using (of which I don't know the provenance) or he found it necessary to add for some reason.

Comment (gthompsnWRF) [reply physics/module_mp_thompson.F90:2208]: All the calculations for ```nu_c``` and ```lamc``` should be internally consistent.  If not, then a bug was made somehow.  I reviewed my WRF code and I've never had a line resetting such as these IF-statements.  I suppose I will have to re-check the math, but nu_c is calculated from ```Nt_c```, then ```lamc``` is calculated and so the computations should be "circular" basically.  The division/multipication by rho could be something that got done as an afterthought and introduced an error.

**Comment by AnningCheng-NOAA** [comment]
I would like to suggest starting a new PR and a new issue to address these
changes, which might take more time than we expected and not really related
to coupling between aerosol and microphysics.  This PR has been in the
queue for about a year and we do not want it to be overtaken by other PRs
again.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90]: @grantfirl @gthompsnWRF OK, I will change it as you suggested.

Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90]: @grantfirl @gthompsnWRF I will change the two IF blocks above this one as well.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:2208]: In the same IF block, the nc(k) = nc1d(k)*rho(k) again. Maybe there is a reason to not to divide by rho. 
Here is the code:
if (qc1d(k) .gt. R1) then
            no_micro = .false.
            rc(k) = qc1d(k)*rho(k)
            **nc(k) = MAX(2., MIN(nc1d(k)*rho(k), Nt_c_max))**
            .........
            nc(k) = MIN( DBLE(Nt_c_max), ccg(1,nu_c)*ocg2(nu_c)*rc(k)   &
                  / am_r*lamc**bm_r)
            if (.NOT. (is_aerosol_aware .or. merra2_aerosol_aware)) nc(k) = Nt_c
else
............

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:3631]: Line 3639, (nc1d(k) + ncten(k)*DT) * rho(k), where nc1d(k) = Nt_c/rho(k), ncten(k) = nc_tendency(k)/rho(k). Feels like the unit is correct for line 3640.

File: physics/module_mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:2208]: I just checked WRFv4.3.2, this line "if (.NOT. is_aerosol_aware) nc(k) = Nt_c" is in WRF too. So I will leave it as is.  If this is a potential problem, we could create an issue in this fork in the future.

Comment (ChunxiZhang-NOAA) [reply physics/module_mp_thompson.F90:2208]: Please refer to the following line in the latest WRF model: 
https://github.com/wrf-model/WRF/blob/master/phys/module_mp_thompson.F#L1730

Comment (gthompsnWRF) [reply physics/module_mp_thompson.F90:2208]: This should not be in WRF.  I don't know who added it, but I have been on a much older version for 2-3 years and don't have it in my code - for the obvious reason that I do not believe it to be necessary and only SLOWS the code down by doing more IF-tests.  Too many cooks in the kitchen in my scheme.

File: physics/mp_thompson.F90
Comment (gthompsnWRF) [reply physics/mp_thompson.F90:881]: @AnningCheng-NOAA   So then PLEASE write a proper description as a comment.  This is not a burden and this is a huge help to those of us who know nothing about MERRA aerosols.  The community needs a proper description.  I will adamantly refuse to accept the PR if one is not included.  It could still be merged by someone without a description, but that would be shame.

**Comment by AnningCheng-NOAA** [comment]
@gthompsnWRF a comment was yesterday added as   "In GOCART and MERRA2,
aerosols are given as mixing ratio (kg/kg). To convert from kg/kg to #/kg,
the "unit mass" (mass of one particle) within the mass bins is calculated.
A lognormal size distribution within aerosol bins is used to find the size
based upon the median mass. NIFA is mainly summarized over five dust bins
and NWFA over the other 10 bins. The parameters besides each bins are
carefully tuned for a good performance of the scheme. "

Please let me know if you have any suggestions.

File: physics/mp_thompson.F90
Comment (ChunxiZhang-NOAA) [reply physics/mp_thompson.F90:881]: @AnningCheng-NOAA @gthompsnWRF The description has been added into the code.

**Comment by gthompsnWRF** [comment]
@AnningCheng-NOAA   I am seeking a bigger description, most importantly, as I indicated at the start, please indicate what are the constituents that represent the last index of the ```aerfld``` array.  In other words, which is organic carbon, sulfates, sea-salts, etc.  And, if any of them are size-bin such as dust, please specify something about the range of the bins.  This will greatly help others, particularly me when anyone else wants to discuss adopting the T-E scheme to another aerosol dataset.  Thank you.

**Comment by AnningCheng-NOAA** [comment]
@chunxi zhang - NOAA Affiliate ***@***.***> , please add the
following description: The fields for the last index of the aerfld array
are specified as below. The units are micrometers.  1: dust bin1, 0.1 to 1;
2: dust bin2, 1 to 1.8; 3: dust bin3, 1.8 to 3.0; 4: dust bin 4, 3.0 to
6.0; 5: dust bin5, 6.0 to 10.0; 6: sea salt bin1, 0.03 to 0.1; 7: sea salt
bin2, 0.1 to 0.5; 8: sea salt bin3, 0.5 to 1.5; 9: sea salt bin4, 1.5 to
5.0; 10: sea salt bin5, 5.0 to 10.0; 11: Sulfate with mean 0.35; 15: water
friendly organic carbon with mean 0.35.
The density has a unit of kg per square meters. dust bin1: 2500; dust
bin2-5: 2650; sea salt all bins: 2200; sulfate: 1700; organic carbon: 1800.
Thank  @gthompsnWRF <https://github.com/gthompsnWRF> for the specific
suggestions. Please let me know if you have any questions.

**Comment by gthompsnWRF** [comment]
@AnningCheng-NOAA  @ChunxiZhang-NOAA   Perhaps this is a problem with Github, but I did not see the new comment from late yesterday added into the code itself.  I even refreshed the ```files changed``` page but it is not showing.

When I directly view the commit tag ```9c1943c``` I do not see the update of documentation by @AnningCheng-NOAA.  I would like to respectfully request that this merge commit be reverted.  Then the comment can be included and then the merge should proceed.  It is really imperative that others know what species are included.

**Comment by ChunxiZhang-NOAA** [comment]
> @AnningCheng-NOAA @ChunxiZhang-NOAA Perhaps this is a problem with Github, but I did not see the new comment from late yesterday added into the code itself. I even refreshed the `files changed` page but it is not showing.

@gthompsnWRF Sorry for that. The PR commit process had started when your latest comment showed up (It took more than one day). We will add the detailed description and combine with other PR.

**Comment by grantfirl** [comment]
> > @AnningCheng-NOAA @ChunxiZhang-NOAA Perhaps this is a problem with Github, but I did not see the new comment from late yesterday added into the code itself. I even refreshed the `files changed` page but it is not showing.
> 
> @gthompsnWRF Sorry for that. The PR commit process had started when your latest comment showed up (It took more than one day). We will add the detailed description and combine with other PR.

@ChunxiZhang-NOAA The additional documentation was added in https://github.com/ufs-community/ccpp-physics/pull/1/commits/9009159abc259bce38c59b1dbc51f8cd54f6544b.

**Comment by ChunxiZhang-NOAA** [comment]
@grantfirl Thank you Grant!

**Pull request merged**
